src/utils/time.ts

```diff
@@ -0,0 +1,93 @@
+export interface TimeComponents {
+    hours?:number;
+    minutes?:number;
+    seconds?:number;
+    ms?:number;
+}
+
+export interface DateComponents {
+    years?:number,
+    months?:number,
+    days?:number,
+}
+
+export interface DateTimeComponents extends DateComponents, TimeComponents {
+
+}
+
+export function addTimedelta(date: Date, components: DateTimeComponents) {
+    return new Date(
+        date.getFullYear()+(components.years||0),
+        date.getMonth()+(components.months||0),
+        date.getDate()+(components.days||0),
+        date.getHours()+(components.hours||0),
+        date.getMinutes()+(components.minutes||0),
+        date.getSeconds()+(components.seconds||0),
+        date.getMilliseconds()+(components.ms||0)
+    );
+}
+
+export function subtractTimedelta(date: Date, components: DateTimeComponents) {
+    return addTimedelta(date, negateTimeComponents(components));
+}
+
+export function negateTimeComponents(components: DateTimeComponents) {
+    const negComponents:TimeComponents = {}
+    for (const key in components) {
+        if (components[key]) {
+            negComponents[key] = components[key] * -1;
+        }
+    }
+    return negComponents;
+}
+
+/**
+ * Normalizes the time of a given date to full divisors of the components.
+ * Example:
+ *  hours: 4 -> 0, 4, 8, 12, 16, 20, 24
+ *  minutes: 5 -> 0, 5, 10, 15, 20 ...
+ *  seconds: 0 -> 0
+ *  date: 13:44:33 -> 12:40:00
+ */
+export function normalizeTime(date: Date, components: TimeComponents) {
```

Comment:
This function, as far as I can see, is the actual raison d'être for this PR, is it not?

I'd like if we could try and mirror the layout of [Temporal](https://tc39.es/proposal-temporal/docs/index.html), for ease of eventual removal of the other functions, for which equivalents can be found in any currently available time manipulation library - if not in any inter-operable form.

... It _is_ fine as it currently is, but if you could spare the time for this adaption, that would be great. If not, just give a holler, and I'll merge it as-is.

Reply:
We could use and adapt Temporals as soon as it's final. This approach just wants to provide interfaces based on Javascript vanilla date features without using another "huge" library like moment or other suggestions here (https://momentjs.com/docs/#/-project-status/recommendations/)

Reply:
Yes, well, it will be easier to deprecate (and for users to act on the deprecation) once whatever final state of Temporal arrives, the closer it mirrors that one's structure. E.g. "in a perfect world" they'd just have to change imports and be done with it.

... I am going to interpret this as "won't do" though, so I'll go ahead and merge this PR ~16.00 UTC+2 (in case you do tell me to wait).

Reply:
No I'm fine.